ledger/complete/ledger.go

```diff
@@ -229,10 +236,39 @@ func (l *Ledger) Set(update *ledger.Update) (newState ledger.State, trieUpdate *
 
 	state := update.State()
 	l.logger.Info().Hex("from", state[:]).
-		Hex("to", newRootHash[:]).
+		Hex("to", newState[:]).
 		Int("update_size", update.Size()).
 		Msg("ledger updated")
-	return ledger.State(newRootHash), trieUpdate, nil
+	return newState, trieUpdate, nil
+}
+
+func (l *Ledger) set(trieUpdate *ledger.TrieUpdate) (newState ledger.State, err error) {
+
+	resultCh := make(chan error, 1)
+
+	trieCh := make(chan *trie.MTrie, 1)
+	defer close(trieCh)
+
+	l.trieUpdateCh <- &WALTrieUpdate{Update: trieUpdate, ResultCh: resultCh, TrieCh: trieCh}
```

Comment:
```suggestion
	// the following two processes are running concurrently:     
	// 1. writing the trie update to WAL 
	// 2. creating a new root node for the trie update
	// since process 1 happens in a different thread, we use the resultCh to wait for it to finish.
	// even though these two processes happens concurrently, process 1 actually depends
	//  on the result of process 2.
	// since the compactor caches the root nodes for minimize memory foot-print while checkpointing,
	// process 1 needs the root trie node from the result of process 2, that's why we create a `trieCh` for
	// process 2 to pass the created trie node to process 1. 
	l.trieUpdateCh <- &WALTrieUpdate{Update: trieUpdate, ResultCh: resultCh, TrieCh: trieCh}
```